nomenclature/code.py

```diff
@@ -19,47 +35,102 @@ def from_dict(cls, mapping):
+            },
+        )
 
     def set_attribute(self, key, value):
         self.attributes[key] = value
 
-
-class Tag(Code):
-    """A simple class for a mapping of a "{tag}" to "target codes" and attributes"""
-
-    attributes: List[
-        Dict[str, Union[str, Dict[str, Union[str, int, float, bool, List, None]], None]]
-    ]
-
-
-def replace_tags(code_list, tag, tag_dict):
-    """Replace tags in `code_list` by `tag_dict`"""
-
-    _code_list = []
-
-    for code in code_list:
-        if f"{{{tag}}}" in code.name:
-            _code_list.extend(_replace_tags(code, tag, tag_dict))
+    @classmethod
+    def named_attributes(cls) -> List[str]:
+        return [a for a in cls.__dict__["__fields__"].keys() if a != "attributes"]
+
+    @property
+    def contains_tags(self) -> bool:
+        return re.search("{.*}", self.name) is not None
+
+    @property
+    def tags(self):
+        return re.findall("(?<={).*?(?=})", self.name)
+
+    def replace_tag(self, tag: str, target):
+        """Return a new instance with tag applied
+
+        Parameters
+        ----------
+        tag : str
+            Name of the tag
+        target : _type_
+            Code that is inserted
+
+        Returns
+        -------
+        _type_
+            new instance with occurrences of "{tag}" replaced by target
+        """
+
+        mapping = {
+            key: value for key, value in self.dict().items() if key != "attributes"
+        }
+        mapping["name"] = mapping["name"].replace("{" + tag + "}", target.name)
+        mapping["description"] = mapping["description"].replace(
+            "{" + tag + "}", target.description
+        )
+        attributes = self.attributes.copy()
+        for attr, value in target.attributes.items():
+            if attr in attributes:
+                attributes[attr] = attributes[attr].replace("{" + tag + "}", value)
+
+        return self.__class__(**mapping, attributes=attributes)
+
+    def __getattr__(self, k):
+        try:
+            return self.attributes[k]
+        except KeyError as ke:
+            raise AttributeError from ke
+
+    def __setattr__(self, name, value):
+        if name not in self.__class__.named_attributes():
+            self.attributes[name] = value
         else:
-            _code_list.append(code)
-
-    return _code_list
-
-
-def _replace_tags(code, tag, target_list):
-    """Utility implementation to replace tags in each item and update attributes"""
-
-    _code_list = []
+            super().__setattr__(name, value)
 
-    for target in target_list:
-        key = code.name.replace(f"{{{tag}}}", target.name)
-        attrs = code.attributes.copy()
-        for _key, _value in target.attributes.items():
-            if _key in attrs:
-                attrs[_key] = attrs[_key].replace(f"{{{tag}}}", _value)
 
-        _code = Code(name=key, attributes=attrs)
-        _code_list.append(_code)
+class VariableCode(Code):
 
-    return _code_list
+    unit: Optional[str]
```

Comment:
This should be required! Even if the variable is dimensionless, the current logic is that a unit attribute has to be explicitly stated in the yaml, e.g.
```yaml
- Some Variable:
    description: some text
    unit: 
```

Reply:
Made it mandatory as described here: https://pydantic-docs.helpmanual.io/usage/models/#required-optional-fields.